File: chapter_2/Q2_2.cpp
/**
 * Tyler Spring
 * 10/27/2024
 * Chapter 2 Return kth to last: Implement an algorithm to find the kth to last
 * element of a singly linked list.
 */
#include <iostream>
using namespace std;

// Define the ListNode structure
struct ListNode {
    int data;
    ListNode* next;

    ListNode(int val) : data(val), next(nullptr) {}
};

class Index {
public:
    int value = 0;
};

// Recursive function to find the kth-to-last element
ListNode* kthToLastHelper(ListNode* head, int k, Index& idx) {
    if (head == nullptr) {
        return nullptr;
    }

    ListNode* node = kthToLastHelper(head->next, k, idx);
    idx.value++;

    if (idx.value == k) {
        return head;
    }
    return node;
}

// Wrapper function for cleaner recursion
ListNode* findKthToLastRecursive(ListNode* head, int k) {
    Index idx;
    return kthToLastHelper(head, k, idx);
}

// Function to print the list (for testing)
void printList(ListNode* head) {
    ListNode* current = head;
    while (current != nullptr) {
        cout << current->data << " -> ";
        current = current->next;
    }
    cout << "null" << endl;
}

int main() {
    // Create a sample linked list: 1 -> 2 -> 3 -> 4 -> 5 -> null
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);

    cout << "Original List: ";
    printList(head);

    int k = 2;
    ListNode* res = findKthToLastRecursive(head, k);

    if (res != nullptr) {
        cout << "Kth to last element is: " << res->data << endl;
    } else {
        cout << "List is too short." << endl;
    }

    // Clean up
    ListNode* temp;
    while (head != nullptr) {
        temp = head;
        head = head->next;
        delete temp;
    }

    return 0;
}
/**
 * Explanation:
 * The function traverses to the end of the list using recursion
 * As the call stack unwinds, we increment the index
 * When the index matches k, we return that node
 * 
 * Index class:
 * we use an Index object to maintain a shared counter across recursive calls
 * 
 * Memory cleanup:
 * we manually free the allocated memory at the end to avoid memory leaks.
 */

File: chapter_2/Q2_2.java
/**
 * Tyler Spring
 * 10/27/2024
 * Chapter 2 Return kth to last: Implement an algorithm to find the kth to last
 * element of a singly linked list.
 */
public class Q2_2 {
public static void main(String[] args) {
    ListNode head = new ListNode(1);
    head.next = new ListNode(2);
    head.next.next = new ListNode(3);
    head.next.next.next = new ListNode(9);
    head.next.next.next.next = new ListNode(5);

    int k = 2;
    ListNode res = findKthToLast(head, k);
    if (res != null) {
        System.out.println("Kth to last element is: " + res.data);
    } else {
        System.out.println("List is too short");
    }
}

//SOLUTION ONE

public static ListNode findKthToLast(ListNode head, int k){
    ListNode p1 = head;
    ListNode p2 = head;

    //Move p2 k nodes ahead
    for (int i = 0; i < k; i++) {
        if (p2 == null) return null;
        p2 = p2.next;
    }

    //Move both pointers until p2 reaches the end
    while (p2 != null) {
        p1 = p1.next;
        p2 = p2.next;
    }
    return p1;
}
    

//SOLUTION TWO

    public static ListNode findKthToLastByLength(ListNode head, int k) {
        int length = 0;
        ListNode current = head;

        //First pass: calculate the length of the List
        while (current != null) {
            length++;
            current = current.next;
        }

        if (k > length) return null; //If k is greater than the length
        
        //Second pass: Find (length - k)th node
        current = head;
        for (int i = 0; i < length -k; i++) {
            current = current.next;
        }
        return current;
    }
}
/**
 * SOLUTION ONE Explanation:
 * p2 is advanced k steps ahead of p1.
 * Both p1 and p2 move together until p2 reaches the end.
 * At this point, p1 will be at the kth-to-last node.
 * 
 * 
 * SOLUTION TWO Explanation:
 * First we compute the length of the List
 * In the second pass, we move to the (length - k)th node, which is the kth-to-last node.
 */